src/device.rs

```diff
@@ -4,12 +4,86 @@
 //
 // SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
 
+use super::queue::Queue;
 use super::*;
-use std::sync::atomic::AtomicUsize;
 use std::sync::Arc;
+use vm_device::interrupt::{InterruptIndex, InterruptSourceGroup, InterruptSourceType};
+use vm_device::resources::DeviceResourceRequest;
 use vm_memory::GuestMemory;
 use vmm_sys_util::eventfd::EventFd;
 
+/// Data struct to manage an virtio queue.
+pub struct VirtioQueueConfig<'a, M: GuestMemory> {
+    /// A virtque.
+    pub queue: Queue<'a, M>,
+
+    /// And associated event notification fd.
+    pub queue_evt: EventFd,
+}
+
+/// Configuration information for an virtio device.
+///
+/// The ownership of an virtio device configuration object will be moved into the VirtioDevice
+/// object when VirtioDevice::activate() method gets called, and it should be returned back
+/// if VirtioDevice::activate() fails or when VirtioDevice::reset() gets called.
+pub struct VirtioDeviceConfig<'a, M: GuestMemory> {
+    /// Memory object to access the guest memory.
+    pub mem: M,
+
+    /// Virtqueues of the virtio device.
+    pub queues: Vec<VirtioQueueConfig<'a, M>>,
+
+    /// A group of interrupts to notify the guest for queue events.
+    pub intr_group: Arc<Box<dyn InterruptSourceGroup>>,
+}
+
+impl<'a, M> VirtioDeviceConfig<'a, M>
+where
+    M: GuestMemory,
+{
+    /// Constructs an virtio device configuration object.
+    pub fn new(
+        mem: M,
+        queues: Vec<VirtioQueueConfig<'a, M>>,
+        intr_group: Arc<Box<dyn InterruptSourceGroup>>,
+    ) -> Self {
+        VirtioDeviceConfig {
+            mem,
+            queues,
+            intr_group,
+        }
+    }
+
+    /// Trigger an interrupt for a specific event source.
+    pub fn trigger(&self, index: u32, flags: u32) -> Result<(), std::io::Error> {
+        match self.intr_group.get_type() {
+            InterruptSourceType::LegacyIrq => self.intr_group.trigger(0, flags),
+            InterruptSourceType::MsiIrq | InterruptSourceType::VfioMsiIrq(_, _) => {
+                self.intr_group.trigger(index, 0)
+            }
+        }
+    }
+
+    /// Get all irqfds associated with vritio queues.
+    pub fn get_vring_irqfds(&self) -> Vec<&EventFd> {
```

Comment:
Is there a usecase where both LegacyIrq and MsiIrq are supported at the same time? It looks like the implementation of the DeviceConfig is wildly different depending on the interrupt system. Each function here is converted in a match on the type of interrupt.

Reply:
Per PCI spec, a device will work in Legacy mode after reset, and may switch to/from PCI MSI/MSI-x mode when OS programs the interrupt configuration registers. So it's a must for a PCI device to support both legacy and MSI/MSIx interrupts.

Reply:
In this case, can we have `VirtioDeviceConfig` as a trait as well? What we have here can be the implementation of the trait and in Firecracker's case we can implement our own Device Configuration which should be much simpler.